### PR TITLE
feat: add support for Twilio Verify

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

Aims to add Twilio Verify Support. Twilio Verify is implemented as a separate provider. Only one of Twilio Verify or Twilio Programmable messaging an be selected. At this time, we only support the use of the `whatsapp` and `sms` channels with Twilio Verify.

This will affect the:
1. Signup flow
2. Verification flow (sms and phone_change)
3. Resend

The token is still generated, but not used in the Twilio Verify flow. It is used as a placeholder so as to try to ensure that to the OTP returned by the Verify service can only be used with the corresponding flow it was generated for. 

## What is the current behaviour?

We support programmable messaging.

## What is the new behaviour?

Developer can toggle between using Twilio Programmable Messaging on all flows or Twilio Verify on all flows.

## Additional context

Manual tests:

Probably need to be conducted on both Phone Change and SMS OTP Verification:

- [x] Existing Programmable Messaging (SMS/WhatsApp) (Signup/Verify/PhoneChange)
- [x] Twilio Verify(SMS/WhatsApp) 
- [ ] Update Frontend to include Twilio Verify

Admin methods shouldn't need to be updated to send to Twilio Verify since admin methods don't require confirmation

### CODE REVIEW

File: internal/api/sms_provider/sms_provider.go [hunk 1]
```diff
@@ -27,6 +29,8 @@ func init() {
 
 type SmsProvider interface {
 	SendMessage(phone, message, channel string) error
+	SendVerification(phone, channel string) error
+	VerifyOTP(phone, otp string) error
```
Comment (kangmingtay) [inline internal/api/sms_provider/sms_provider.go:33]: is there a way to not include this in the interface since it only applies to twilio? we could also create a default `SendVerification` method for the `SmsProvider` which can be overridden by twilio's `SendVerification` method - that should help to cut down some LOC in `internal/api/phone.go`

File: internal/api/sms_provider/twilio.go [hunk 2]
```diff
@@ -107,5 +132,78 @@ func (t *TwilioProvider) SendSms(phone, message, channel string) error {
...
+	if err != nil {
+		return err
+	}
+	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	r.SetBasicAuth(t.Config.AccountSid, t.Config.AuthToken)
+	res, err := client.Do(r)
+	defer utilities.SafeClose(res.Body)
+	if err != nil {
+		return err
+	}
```
Comment (hf) [inline internal/api/sms_provider/twilio.go:190]:
```suggestion
	if err != nil {
		return err
	}
	defer utilities.SafeClose(res.Body)
```

File: internal/api/sms_provider/twilio.go [hunk 3]
```diff
@@ -107,5 +132,78 @@ func (t *TwilioProvider) SendSms(phone, message, channel string) error {
...
+
+func (t *TwilioProvider) SendVerification(phone, channel string) error {
+	// Unlike Programmable Messaging, Verify does not require a prefix for channel
+	// E164 format is also guaranteed by the time this function is called
+	body := url.Values{
+		"To":      {phone},
+		"Channel": {channel},
+	}
+	client := &http.Client{Timeout: defaultTimeout}
+	r, err := http.NewRequest("POST", t.APIPath, strings.NewReader(body.Encode()))
```
Comment (hf) [inline internal/api/sms_provider/twilio.go:147]:
```suggestion
	r, err := http.NewRequest(http.MethodPost, t.APIPath, strings.NewReader(body.Encode()))
```

File: internal/api/verify.go [hunk 4]
```diff
@@ -494,6 +495,14 @@ func (a *API) verifyUserAndToken(ctx context.Context, conn *storage.Connection,
...
 		case smsVerification:
 			user, err = models.FindUserByPhoneAndAudience(conn, params.Phone, aud)
+			smsProvider, _ := sms_provider.GetSmsProvider(*config)
+			if config.Sms.Provider == "twilio" && config.Sms.Twilio.VerifyEnabled {
+				if err := smsProvider.VerifyOTP(params.Phone, params.Token); err != nil {
+					return nil, expiredTokenError("Token has expired or is invalid").WithInternalError(err)
+				}
+				return user, nil
+			}
+
```
Comment (hf) [inline internal/api/verify.go:505]: Why is this necessary? Shouldn't it be exactly the same as the existing providers?

File: internal/conf/configuration.go [hunk 5]
```diff
@@ -205,6 +205,7 @@ type SmsProviderConfiguration struct {
 }
 
 type TwilioProviderConfiguration struct {
+	VerifyEnabled     bool   `json:"verify_enabled" split_words:"true"`
 	AccountSid        string `json:"account_sid" split_words:"true"`
 	AuthToken         string `json:"auth_token" split_words:"true"`
 	MessageServiceSid string `json:"message_service_sid" split_words:"true"`
```
Comment (hf) [inline internal/conf/configuration.go:211]: In my opinion, you should extract Twilio Verifty to be a _new_ provider. So register a new struct `TwilioVerifyProvider` etc. Don't re-use the existing one, which you can rename `TwilioProgrammableMessagingProvider` or such.

File: internal/api/verify.go (hunk 4 above)
Comment (J0) [reply internal/api/verify.go:505]: Yup, you are right this section verifies the user and should be the same as existing providers. We still need to  verify the token with Twilio's Verification Service and that would probably make more sense if done before the `isOtpValid` check.

Let me know if there's anything I'm missing.

File: internal/api/sms_provider/sms_provider.go [hunk 6]
```diff
@@ -14,6 +14,8 @@ var defaultTimeout time.Duration = time.Second * 10
 const SMSProvider = "sms"
 const WhatsappProvider = "whatsapp"
 
+const VerifyUnsupportedError = "external otp verification is not supported for this provider"
```
Comment (J0) [inline internal/api/sms_provider/sms_provider.go:17]:
```suggestion
```

File: internal/api/sms_provider/twilio_verify.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	switch channel {
+	case SMSProvider, WhatsappProvider:
+		return "", t.SendSms(phone, message, channel)
+	default:
+		return "", fmt.Errorf("channel type %q is not supported for Twilio", channel)
+	}
+}
+
+// Send an SMS containing the OTP with Twilio's API
+func (t *TwilioVerifyProvider) SendSms(phone, message, channel string) error {
```
Comment (hf) [inline internal/api/sms_provider/twilio_verify.go:65]: Pretty sure there's a message ID to be returned here?

Comment (J0) [reply internal/api/sms_provider/twilio_verify.go:65]: Hm I don't think there's a MessageID for Twilio Verify. 

Unless you are referring to the Verification ID? https://www.twilio.com/docs/verify/api/verification?code-sample=code-start-a-verification-with-sms&code-language=curl&code-sdk-version=json#verification-response-properties

Comment (J0) [reply internal/api/sms_provider/twilio_verify.go:65]: (Wondering because Verifications [don't give an indication of Delivery unlike the MessageID](https://www.twilio.com/docs/verify/api/verification) not sure how useful it will be in terms of debugging)

**Review by kangmingtay — APPROVED** [review]
nice work, this looks great! should be fine to merge if you've tested it with twilio verify end to end too
